File: frontend/src/lib/api/customer.ts
/**
 * ============================================================
 * CUSTOMER SERVICE - Authenticated Customer API Calls
 * ============================================================
 * Balance, Transactions, TopUp, Profile for logged-in users.
 * ============================================================
 */

import { api, type ApiResponse, type PaginatedData } from "./client";
import { ENDPOINTS } from "./config";

// ----------------------------------------------------------
// Types - Balance
// ----------------------------------------------------------

export interface BalanceData {
  balance: number;
  bonus_balance: number;
  available_balance: number;
}

export interface BalanceMutation {
  id: number;
  type: "credit" | "debit";
  amount: number;
  balance_before: number;
  balance_after: number;
  description: string;
  created_at: string;
  transaction?: {
    transaction_code: string;
  };
}

// ----------------------------------------------------------
// Types - Customer Transactions
// ----------------------------------------------------------

export interface CustomerTransaction {
  id: number;
  transaction_code: string;
  transaction_type: "purchase" | "topup";
  product_name: string;
  product_code?: string;
  product_price: number;
  admin_fee: number;
  discount_amount: number;
  total_price: number;
  customer_data?: Record<string, string>;
  payment_method: string;
  status: string;
  payment_status: string;
  created_at: string;
  paid_at?: string;
  completed_at?: string;
  sn?: string | null;
  note?: string | null;
  product?: {
    id: number;
    name: string;
    image_url?: string;
  };
  payment?: {
    payment_code?: string;
    payment_method: string;
    payment_channel?: string;
    amount: number;
    status?: string;
    payment_url?: string;
    qr_code_url?: string;
    expired_at?: string;
    instructions?: Array<{ title: string; steps: string[] }>;
  };
  logs?: Array<{
    action: string;
    description: string;
    created_at: string;
  }>;
}

export interface TransactionFilters {
  type?: "purchase" | "topup";
  status?: string;
  search?: string;
  start_date?: string;
  end_date?: string;
  per_page?: number;
  page?: number;
}

// ----------------------------------------------------------
// Types - TopUp Request
// ----------------------------------------------------------

export interface TopUpRequest {
  amount: number;
  payment_method: string;
}

export interface TopUpResponse {
  transaction: {
    id: number;
    transaction_code: string;
    transaction_type: "topup";
    product_name: string;
    product_price: number;
    admin_fee: number;
    total_price: number;
    status: string;
    payment_status: string;
    expired_at: string;
  };
  payment: {
    payment_code: string;
    payment_method: string;
    payment_channel: string;
    amount: number;
    payment_url: string;
    qr_code_url?: string;
    expired_at: string;
  };
}

// ----------------------------------------------------------
// Types - Purchase via Balance
// ----------------------------------------------------------

export interface PurchaseBalanceRequest {
  product_item_id: number;
  customer_data: Record<string, string>;
  voucher_code?: string;
}

// ----------------------------------------------------------
// Types - Profile
// ----------------------------------------------------------

export interface CustomerProfile {
  id: number;
  name: string;
  email: string;
  phone: string;
  role: string;
  customer_type: "retail" | "reseller";
  status: string;
  avatar?: string;
  created_at: string;
  balance?: {
    balance: number;
    bonus_balance: number;
  };
}

// ----------------------------------------------------------
// Types - Postpaid
// ----------------------------------------------------------

export interface PostpaidInquiryRequest {
  product_item_id: number;
  customer_no: string;
}

export interface PostpaidBillDetail {
  period: string;
  nominal: number;
  admin: number;
  denda: number;
  meter_awal?: string | null;
  meter_akhir?: string | null;
  biaya_lain?: number;
}

export interface PostpaidInquiryData {
  inquiry_ref: string;
  product_name: string;
  customer_no: string;
  customer_name: string;
  period: string;
  tagihan: number;
  admin: number;
  total: number;
  details: PostpaidBillDetail[];
}

export interface PostpaidInquiryResponse {
  data: PostpaidInquiryData;
}

export interface PostpaidPayRequest {
  inquiry_ref: string;
  payment_method: string;
}

// ----------------------------------------------------------
// Balance API
// ----------------------------------------------------------

export async function getBalance(): Promise<ApiResponse<BalanceData>> {
  return api.get<BalanceData>(ENDPOINTS.customer.balance, undefined, true);
}

export async function getBalanceMutations(
  limit = 50,
): Promise<ApiResponse<BalanceMutation[]>> {
  return api.get<BalanceMutation[]>(
    ENDPOINTS.customer.balanceMutations,
    { limit },
    true,
  );
}

// ----------------------------------------------------------
// Transaction API (Authenticated)
// ----------------------------------------------------------

export async function getTransactions(
  filters?: TransactionFilters,
): Promise<ApiResponse<PaginatedData<CustomerTransaction>>> {
  return api.get<PaginatedData<CustomerTransaction>>(
    ENDPOINTS.customer.transactions,
    filters as any,
    true,
  );
}

export async function getTransactionDetail(
  code: string,
): Promise<ApiResponse<CustomerTransaction>> {
  return api.get<CustomerTransaction>(
    ENDPOINTS.customer.transactionDetail(code),
    undefined,
    true,
  );
}

export async function customerPurchase(data: {
  product_item_id: number;
  payment_method: string;
  customer_data: Record<string, string>;
  voucher_code?: string;
}): Promise<ApiResponse<any>> {
  return api.post(ENDPOINTS.customer.purchase, data, true);
}

export async function purchaseWithBalance(
  data: PurchaseBalanceRequest,
): Promise<ApiResponse<any>> {
  return api.post(ENDPOINTS.customer.purchaseBalance, data, true);
}

// ----------------------------------------------------------
// TopUp API
// ----------------------------------------------------------

export async function createTopUp(
  data: TopUpRequest,
): Promise<ApiResponse<TopUpResponse>> {
  return api.post<TopUpResponse>(ENDPOINTS.customer.topup, data, true);
}

export async function getCustomerPaymentChannels(): Promise<
  ApiResponse<any[]>
> {
  return api.get<any[]>(ENDPOINTS.customer.paymentChannels, undefined, true);
}

// ----------------------------------------------------------
// Postpaid API
// ----------------------------------------------------------

export async function postpaidInquiry(
  data: PostpaidInquiryRequest,
): Promise<ApiResponse<PostpaidInquiryData>> {
  return api.post<PostpaidInquiryData>(
    ENDPOINTS.customer.postpaidInquiry,
    data,
    true,
  );
}

export async function postpaidPay(
  data: PostpaidPayRequest,
): Promise<ApiResponse<any>> {
  return api.post<any>(ENDPOINTS.customer.postpaidPay, data, true);
}

// ----------------------------------------------------------
// Profile API
// ----------------------------------------------------------

export async function getProfile(): Promise<ApiResponse<CustomerProfile>> {
  return api.get<CustomerProfile>(ENDPOINTS.customer.profile, undefined, true);
}

export async function updateProfile(
  data: Partial<Pick<CustomerProfile, "name" | "email" | "phone" | "avatar">>,
): Promise<ApiResponse<CustomerProfile>> {
  return api.put<CustomerProfile>(ENDPOINTS.customer.updateProfile, data, true);
}

export async function changePassword(data: {
  current_password: string;
  new_password: string;
  new_password_confirmation: string;
}): Promise<ApiResponse<void>> {
  return api.post<void>(ENDPOINTS.customer.changePassword, data, true);
}

// ----------------------------------------------------------
// Notification API
// ----------------------------------------------------------

export interface AppNotification {
  id: string;
  title: string;
  message: string;
  type: string;
  data: {
    action_url?: string;
    icon?: string;
    color?: string;
    [key: string]: any;
  };
  read_at: string | null;
  created_at: string;
  updated_at: string;
}

export async function getNotifications(
  page = 1,
): Promise<ApiResponse<PaginatedData<AppNotification>>> {
  return api.get<PaginatedData<AppNotification>>(
    ENDPOINTS.customer.notifications.list,
    { page },
    true,
  );
}

export async function getUnreadNotificationCount(): Promise<
  ApiResponse<{ count: number }>
> {
  return api.get<{ count: number }>(
    ENDPOINTS.customer.notifications.count,
    undefined,
    true,
  );
}

export async function markNotificationAsRead(
  id: string,
): Promise<ApiResponse<void>> {
  return api.post<void>(
    ENDPOINTS.customer.notifications.markAsRead(id),
    undefined,
    true,
  );
}

export async function markAllNotificationsAsRead(): Promise<ApiResponse<void>> {
  return api.post<void>(
    ENDPOINTS.customer.notifications.markAllAsRead,
    undefined,
    true,
  );
}

File: enstore/lib/frontend/api/customer.dart
import 'client.dart';
import 'config.dart';
import 'types.dart';

class CustomerService {
  static final CustomerService _instance = CustomerService._internal();

  factory CustomerService() {
    return _instance;
  }

  CustomerService._internal();

  Future<ApiResponse<CustomerProfile>> getProfile() async {
    return await ApiClient().get<CustomerProfile>(
      ApiConfig.endpoints.customer.profile,
      fromJson: (data) => CustomerProfile.fromJson(data),
    );
  }

  Future<ApiResponse<CustomerProfile>> updateProfile(
    Map<String, dynamic> data,
  ) async {
    return await ApiClient().put<CustomerProfile>(
      ApiConfig.endpoints.customer.updateProfile,
      data: data,
      fromJson: (data) => CustomerProfile.fromJson(data),
    );
  }

  Future<ApiResponse<void>> changePassword(
    String currentPassword,
    String newPassword,
    String confirmPassword,
  ) async {
    return await ApiClient().post<void>(
      ApiConfig.endpoints.customer.changePassword,
      data: {
        'current_password': currentPassword,
        'new_password': newPassword,
        'new_password_confirmation': confirmPassword,
      },
    );
  }

  Future<ApiResponse<BalanceData>> getBalance() async {
    return await ApiClient().get<BalanceData>(
      ApiConfig.endpoints.customer.balance,
      fromJson: (data) => BalanceData.fromJson(data),
    );
  }

  Future<ApiResponse<List<BalanceMutation>>> getBalanceMutations({
    int limit = 50,
  }) async {
    return await ApiClient().get<List<BalanceMutation>>(
      ApiConfig.endpoints.customer.balanceMutations,
      queryParameters: {'limit': limit},
      fromJson: (data) =>
          (data as List).map((e) => BalanceMutation.fromJson(e)).toList(),
      // Fix: Types in ApiClient.get might need adjustment for List<T> or wrapper
    );
  }

  Future<ApiResponse<PaginatedData<CustomerTransaction>>> getTransactions({
    Map<String, dynamic>? filters,
  }) async {
    return await ApiClient().get<PaginatedData<CustomerTransaction>>(
      ApiConfig.endpoints.customer.transactions,
      queryParameters: filters,
      fromJson: (data) => PaginatedData.fromJson(
        data,
        (item) => CustomerTransaction.fromJson(item),
      ),
    );
  }

  Future<ApiResponse<CustomerTransaction>> getTransactionDetail(
    String code,
  ) async {
    return await ApiClient().get<CustomerTransaction>(
      ApiConfig.endpoints.customer.transactionDetail(code),
      fromJson: (data) => CustomerTransaction.fromJson(data),
    );
  }

  // Notifications
  Future<ApiResponse<PaginatedData<AppNotification>>> getNotifications({
    int page = 1,
  }) async {
    return await ApiClient().get<PaginatedData<AppNotification>>(
      ApiConfig.endpoints.customer.notifications.list,
      queryParameters: {'page': page},
      fromJson: (data) => PaginatedData.fromJson(
        data,
        (item) => AppNotification.fromJson(item),
      ),
    );
  }

  Future<ApiResponse<int>> getUnreadNotificationCount() async {
    return await ApiClient().get<int>(
      ApiConfig.endpoints.customer.notifications.count,
      fromJson: (data) =>
          data['count'] ?? 0, // Assuming { count: 5 } response format
    );
  }

  Future<ApiResponse<void>> markNotificationAsRead(String id) async {
    return await ApiClient().post<void>(
      ApiConfig.endpoints.customer.notifications.markAsRead(id),
    );
  }

  Future<ApiResponse<void>> markAllNotificationsAsRead() async {
    return await ApiClient().post<void>(
      ApiConfig.endpoints.customer.notifications.markAllAsRead,
    );
  }
}
